Adopt CompactVariant in Style::LengthPercentage/Style::AnglePercentage

weinig opened the pull request:
#### c5606cb19ea924582d75939ec6343cec295f1d7e
<pre>
Adopt CompactVariant in Style::LengthPercentage/Style::AnglePercentage
<a href="https://bugs.webkit.org/show_bug.cgi?id=285250">https://bugs.webkit.org/show_bug.cgi?id=285250</a>

Reviewed by Darin Adler.

Replaces hand rolled compact variant, `PrimitiveDimensionPercentage`,
used by `Style::LengthPercentage` and `Style::AnglePercentage` with
the new `WTF::CompactVariant` type.

To aid this, the direct use of `Ref&lt;CalculationValue&gt;` was replaced
with a new generic `Style::UnevaluatedCalculation&lt;&gt;` which serves as
the Style layer analog to the CSS layers `CSS::UnevaluatedCalc&lt;&gt;`.
Mark both types as being smart pointers by specializing `IsSmartPtr`
so that they both can be used in `CompactVariants`.

`Style::LengthPercentage` and `Style::AnglePercentage` were almost
identical, so rather than updating both of them to use the new type,
the generic `DimensionPercentage&lt;&gt;` was added, allowing subclassing
to define `Style::LengthPercentage` and `Style::AnglePercentage`.

A few issues arose with `WTF::CompactVariant` that ended up being
due to the implementations of move and copy not encoding types with
alternative representations correctly. This has been corrected and
tests have been added.

Additionally, it was useful to be able to swap in `std::variant&lt;&gt;`
in place of `WTF::CompactVariant` to debug issues. To make this
feasible without changing code, a new WTF::holdsAlternative&lt;&gt; was
added that works with any &quot;variant-like&quot;.

* Source/WTF/wtf/CompactVariant.h:
* Source/WTF/wtf/CompactVariantOperations.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Calculation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h: Added.
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:

Canonical link: <a href="https://commits.webkit.org/288346@main">https://commits.webkit.org/288346@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e462512c20619484c48c588cdbd3c7c708fa36b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82669 "29 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36823 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87789 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33727 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64414 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22182 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85726 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75200 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44689 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1636 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29386 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32761 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75654 "Found 1 new JSC stress test failure: stress/json-raw-json.js.ftl-eager-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72913 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30045 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89155 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81717 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9962 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7215 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72043 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17912 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16194 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1351 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9915 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15436 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104123 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9789 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25244 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/CompactVariant.h:
```diff
@@ -33,29 +33,33 @@ namespace WTF {
-        requires std::constructible_from<V, U> && (!std::same_as<std::remove_cvref_t<U>, CompactVariant>)
+        requires
+            std::constructible_from<StdVariant, U>
+         && (!std::same_as<std::remove_cvref_t<U>, CompactVariant>)
+         && (!IsStdInPlaceTypeV<std::remove_cvref_t<U>>)
+         && (!IsStdInPlaceIndexV<std::remove_cvref_t<U>>)
```
Why the parentheses?

weinig replied:
Weird requirement of the requires syntax.

darinadler commented on Source/WTF/wtf/CompactVariant.h:
```diff
@@ -110,6 +118,11 @@ template<CompactVariantCapable V> class CompactVariant {
         Operations::destruct(m_data);
     }
 
+    void swap(CompactVariant& other)
+    {
+        std::swap(m_data, other.m_data);
+    }
```
Do we also want an overload of `std::swap` that does this or a free function `WTF::swap`? Trying to remember the best practice.

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
```diff
@@ -188,7 +188,7 @@ template<auto R, typename... Rest> LengthPercentage<R> canonicalize(const CSS::L
 
 // Out of line to avoid inclusion of CSSCalcValue.h
-Ref<CSSCalcValue> makeCalc(const CalculationValue&, const RenderStyle&);
+Ref<CSSCalcValue> makeCalc(Ref<CalculationValue>, const RenderStyle&);
```
How do you decide when to make an argument `Ref&&` and when to make it `Ref`? And should we write that into the coding style guidelines?

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
```diff
@@ -197,390 +195,140 @@ template<CSS::Range R = CSS::All> struct Flex {
-    static constexpr auto category = Calculation::Category::LengthPercentage;
+namespace WTF {
+ 
+// Allow primitives numeric types that usually stored their value as a `double` to be
```
stored -> store

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
```diff
@@ -197,390 +195,140 @@ template<CSS::Range R = CSS::All> struct Flex {
+namespace WTF {
+ 
+// Allow primitives numeric types that usually stored their value as a `double` to be
+// used with CompactVariant by using a `float`, rather than `double` representation
```
But why is it OK to lose range and precision? What makes this lossless?

weinig replied:
Nothing makes it lossless, but it matches the existing behavior we have for LengthPercentage types in the WebCore::Length class. We may want to reconsider it at some point, but for now, it is at least consistent.

weinig committed:
Adopt CompactVariant in Style::LengthPercentage/Style::AnglePercentage
https://bugs.webkit.org/show_bug.cgi?id=285250

Reviewed by Darin Adler.

Replaces hand rolled compact variant, `PrimitiveDimensionPercentage`,
used by `Style::LengthPercentage` and `Style::AnglePercentage` with
the new `WTF::CompactVariant` type.

To aid this, the direct use of `Ref<CalculationValue>` was replaced
with a new generic `Style::UnevaluatedCalculation<>` which serves as
the Style layer analog to the CSS layers `CSS::UnevaluatedCalc<>`.
Mark both types as being smart pointers by specializing `IsSmartPtr`
so that they both can be used in `CompactVariants`.

`Style::LengthPercentage` and `Style::AnglePercentage` were almost
identical, so rather than updating both of them to use the new type,
the generic `DimensionPercentage<>` was added, allowing subclassing
to define `Style::LengthPercentage` and `Style::AnglePercentage`.

A few issues arose with `WTF::CompactVariant` that ended up being
due to the implementations of move and copy not encoding types with
alternative representations correctly. This has been corrected and
tests have been added.

Additionally, it was useful to be able to swap in `std::variant<>`
in place of `WTF::CompactVariant` to debug issues. To make this
feasible without changing code, a new WTF::holdsAlternative<> was
added that works with any "variant-like".

* Source/WTF/wtf/CompactVariant.h:
* Source/WTF/wtf/CompactVariantOperations.h:
* Source/WTF/wtf/StdLibExtras.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Calculation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h: Added.
* Tools/TestWebKitAPI/Tests/WTF/CompactVariant.cpp:

Canonical link: https://commits.webkit.org/288346@main
